src/vtquery.cpp

```diff
@@ -268,11 +272,24 @@ struct Worker : Nan::AsyncWorker {
             // query point lng/lat geometry.hpp point (used for distance calculation later on)
             mapbox::geometry::point<double> query_lnglat{data.longitude, data.latitude};
 
-            // for each tile
+            gzip::Decompressor decompressor;
+            std::string uncompressed;
+            std::vector<std::string> buffers;
+            std::vector<std::tuple<vtzero::vector_tile, std::int32_t, std::int32_t, std::int32_t>> tiles;
+            tiles.reserve(data.tiles.size());
             for (auto const& tile_ptr : data.tiles) {
                 TileObject const& tile_obj = *tile_ptr;
-                vtzero::vector_tile tile{tile_obj.data};
-
+                if (gzip::is_compressed(tile_obj.data.data(), tile_obj.data.size())) {
+                    decompressor.decompress(uncompressed, tile_obj.data.data(), tile_obj.data.size());
+                    buffers.emplace_back(std::move(uncompressed));
+                    tiles.emplace_back(vtzero::vector_tile(buffers.back()), tile_obj.z, tile_obj.x, tile_obj.y);
```

Comment:
You've noticed the critical thing here. So, `vtzero` is written with top performance in mind. This means that it cannot/does not/refuses to accept the cost of copying or re-allocation internally. Instead, by design, it expects that the memory you pass it will stay alive for as long as `vtzero` needs to work on it. What this means is that the vtzero calls at https://github.com/mapbox/vtquery/blob/de2e1313f9e707ff765b981238bd9d57d94c0b0c/src/vtquery.cpp#L381-L382 need the buffer passed to the original `vtzero::vector_tile` object to still be alive. Otherwise they will reference dangling pointers internally. While this is dangerous, its by design for top performance. As long as the data passed to vtzero stays in scope (and is controlled outside of vtzero) we can have perfectly correct code and top performance without copies.

So, the `std::vector<std::string> buffers;` is my best attempt to deliver `vtzero` this promise. The `std::vector<std::string> buffers;` is at the correct scope such that the data it holds is perfectly alive and in memory for as long as vtzero needs it. More info at https://github.com/mapbox/vtzero/issues/36 and https://github.com/mapbox/vtzero/commit/a42889ddd507c38a3b5f49a0ea8468a8f4a394a4